### DISCUSSION TITLE
Broken nodeport-proxy brings cluster into unusable state

### DISCUSSION THREAD

**Discussion by almereyda** [kubermatic/kubermatic#10739] [Q&A] [answered]
After my recent contributions here, we finally decided to roll back from Kubermatic, due to a sudden break of the connection between the seed cluster and the user clusters. Since our `charts/monitoring` deployment's Grafana did not pick up the `charts/logging` instance (and the Explore tab wasn't shown), we had to investigate this manually. All Kubermatic components (master/seed) were running in the same KubeOne cluster.

Maybe you want to have a look, and can identify this as reason to step back from defaulting on Konnectivity just yet.

- https://circlesland.fibery.io/Ops/Incidents/2/incident-2-Investigation-86?sharing-key=f5a0ebc7-c841-4cda-b1fe-9347afcafe0d  
  make sure to switch to the wide view in the top right

This just broke from one day to the other, and half a week of investigation in the system and documentation here did not yield any pointers at how to repair this. A Kubermatic cluster that has a broken connection to the user clusters is very instable, but pods continue to run at least. The Konnectivity connection is a SPOF here, and HA considerations might apply. The side effects just ripple through the whole system: Observability becomes unusable, and none of the controllers is able to suggest mitigations, nor to reconcile into a healthy state.

<details>

<summary>Logs that could be provided upon request</summary>

Which is all that we retained:

- 2022-08-09-21-51-seed-controller.log
- 2022-08-09-22-03-nodeport-proxy-envoy-envoy.log
- 2022-08-09-22-16-kubermatic-api.log
- 2022-08-10-17-00-apiserver-konnectivity-server.log
- 2022-08-11-17-20-kubermatic-seed-controller-manager.log

</details>

**Comment by rastislavs** [comment, marked as answer]
Hi @almereyda , thanks a lot for the report.

Unfortunately, looking at the report, I am not sure if Konenctivity is really the one to blame. How did you come to that conclusion?
 - Did everything work until you enabled Konenctivity?
 - Did things start working again after you switched back from Konenctivity to OpenVPN?
 - What errors in particular (on apiserver, konnectivity-server / konnectivity-agent) do you think point to the problem with Konenctivity? Some errors found in your logs ("Receive channel on Connect is stopping" / "could not get frontend client" / "Receive channel on Proxy is stopping") are red herrings - expected during the normal operation of Konenctivity, and in fact their severity has been lowered to info/debug logs in more recent Konenctivity changes.
 - The [known issue with Konenctivity](https://github.com/kubernetes-sigs/apiserver-network-proxy/issues/370) is that konnectivity-server might be running out of memory and being OOM killed after longer operational time, have you seen that in your setup?

That said, I still see some other potential problems in your logs:
 - Errors from the seed controller manager - the top one says it cannot configure aleratmanager. It seems that you enabled the [User Cluster MLA](https://docs.kubermatic.com/kubermatic/v2.20/architecture/monitoring_logging_alerting/user_cluster/) feature in your setup (and it was enabled in your seed with [this seed-level flag](https://docs.kubermatic.com/kubermatic/v2.20/tutorials_howtos/monitoring_logging_alerting/user_cluster/admin_guide/#enabling-mla-stack-in-a-seed)). Have you also installed the MLA backend in the seed before (as described in previous steps of the MLA installation guide) and was it running successfully before turning on the seed-level flag of MLA to enabled? (see [Installing MLA Stack in a Seed Cluster](https://docs.kubermatic.com/kubermatic/v2.20/tutorials_howtos/monitoring_logging_alerting/user_cluster/admin_guide/#installing-mla-stack-in-a-seed-cluster))  
 - It also looks like there were some issues with exposing apiserver via nodeport-proxy (note that Konnectivity is not involved in this direction of communication). What [expose strategy](https://docs.kubermatic.com/kubermatic/v2.20/tutorials_howtos/networking/expose_strategies/) did you use - I guess NodePort? What was the health of the nodeport-proxy overall and was the DNS set up properly?

**Reply by almereyda** [reply]
Thank you for your differing interpretation, and for the upstream link. It's good we're in a discussion here and don't noise-up their log.

My conclusion was merely based on the fact, that there was an error in the seed cluster talking to its user cluster, and I believed Konnectivity to be responsible for that connection. Of course there's also a receiving side, which has to be able to react to connections mediated through Konnectivity, which I apparently left out of consideration.

Yes, DNS was set up properly, and we did not explicitly specify an exposeStrategy, which apparently defaults to `NodePort`.

I was sure I missed something. The cluster is now shut down, so I cannot possibly investigate further. It was working well for a week, with all features you asked for enabled/activated/installed, and also working, but then suddenly started to show these regressions.

As the cluster was installed with Konnectivity enabled, I did not know I could merely switch back, if that was the root cause. But I now thought about setting up the next iteration with OpenVPN, until Konnectivity is declared stable. FYI, we have since moved to a manged instance.

Due to the many connections between the components, also a look at Hubble did not allow me to correlate log events to the visible network connections. Esp. since Cilium only registers the connections that take place, and *not* those that don't. Could be nice to have an overlay here, which shows which application errors out on trying to access which, and mark that connection red. Yet this seems a feature request to (the) upstream(s).

I *know* we were affected by

- https://github.com/kubermatic/kubeone/pull/2204

and I *feel*, because I had read the term *finalizers* often in those errors, that we also suffered from

- https://github.com/kubermatic/kubermatic/pull/10536

until we upgraded to v1.4.6 and v2.20.6 respectively.

- Is there any certificate rotation in place at the nodeport-proxy, which might have broken TLS connectivity?
- Could this eventually have been mitigated by switching to the `LoadBalancer` `exposeStrategy`?

**Reply by rastislavs** [reply]
> Is there any certificate rotation in place at the nodeport-proxy, which might have broken TLS connectivity?

No, I don't think so.

> Could this eventually have been mitigated by switching to the LoadBalancer exposeStrategy?

That is really hard to tell without figuring out the root-cause of the issue. The nodeport-proxy is involved also in case of the LoadBalancer strategy, just slightly in different way (separate proxy running for each user cluster).

**Comment by almereyda** [comment]
Thank you for your time and thoughts!

# 🤷

Let's see next time.
